crates/ruff_server/src/server/api/requests/code_action_resolve.rs

```diff
@@ -0,0 +1,103 @@
+use crate::server::api::LSPResult;
+use crate::server::{client::Notifier, Result};
+use crate::server::{AvailableCodeActions, SupportedCodeActionKind};
+use crate::session::DocumentSnapshot;
+use crate::PositionEncoding;
+use lsp_server::ErrorCode;
+use lsp_types::{self as types, request as req};
+use types::WorkspaceEdit;
+
+pub(crate) struct CodeActionResolve;
+
+impl super::RequestHandler for CodeActionResolve {
+    type RequestType = req::CodeActionResolveRequest;
+}
+
+impl super::BackgroundDocumentRequestHandler for CodeActionResolve {
+    fn document_url(params: &types::CodeAction) -> std::borrow::Cow<types::Url> {
+        let uri: lsp_types::Url = serde_json::from_value(params.data.clone().unwrap_or_default())
+            .expect("code actions should have a URI in their data fields");
+        std::borrow::Cow::Owned(uri)
+    }
+    fn run_with_snapshot(
+        snapshot: DocumentSnapshot,
+        _notifier: Notifier,
+        action: types::CodeAction,
+    ) -> Result<types::CodeAction> {
+        let document = snapshot.document();
+
+        let supported: SupportedCodeActionKind = action
+            .kind
+            .clone()
+            .ok_or(anyhow::anyhow!("No kind was given for code action"))
+            .with_failure_code(ErrorCode::InvalidParams)?
+            .try_into()
+            .map_err(|()| anyhow::anyhow!("Code action was of an invalid kind"))
+            .with_failure_code(ErrorCode::InvalidParams)?;
+        let available_action = supported.makes_available();
+
+        // ensures that only one code action kind was made available
+        debug_assert!(
+            (available_action & (available_action - AvailableCodeActions::all()))
+                == AvailableCodeActions::empty()
+        );
+
+        if available_action == AvailableCodeActions::SOURCE_FIX_ALL {
+            resolve_edit_for_fix_all(
+                action,
```

Comment:
I wanted to avoid duplicating the logic of `makes_available`, since I was concerned about a misalignment in logic if we added new variants of code action kinds to be accounted for. However, I'd be fine with using a match statement like this if we make sure it's exhaustive, so that adding new variants would force us to update the `match` logic.